Added new documentation: Send logs from Nodejs winston to Signoz cloud (#1026)
SigNoz/signoz.io: data/docs/userguide/collecting_nodejs_winston_logs.mdx

---
date: 2024-12-06
title: Collecting Nodejs winston logs
id: collecting_nodejs_winston_logs
---

If you are using `winston` as your logging library in your Nodejs application, you can export these logs to SigNoz very easily using various transports provided by `winston`.

> <Admonition>

Reviewer: ```suggestion

```

Please remove this admonition and add it to the relevant place if needed. If it is not required by a user, please remove it completely.

The default logging level can be configured.

To send all the logs to SigNoz please change the default log level to DEBUG.

```
const logger = winston.createLogger({
  level: 'debug',
  // ... other configurations
});
```

</Admonition>

> <Tabs>

Reviewer: ```suggestion
<Tabs entityName="plans">
```

This is needed for SigNoz SaaS

<TabItem value="cloud" label="SigNoz Cloud" default>

> For sending logs to SigNoz cloud, while running the above example set the below environment variables

Reviewer: ```suggestion
```
Mention these details about the environment variables at the relevant place rather than mentioning them at the start. That way it will be more helpful for the user.

> - The value of `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable will be `https://ingest.{region}.signoz.cloud:443/v1/logs` where depending on the choice of your region for SigNoz cloud, the otlp endpoint will vary according to this table.

Reviewer: ```suggestion
```

>   | Region | Endpoint                   |

Reviewer: ```suggestion
```

>   | ------ | -------------------------- |

Reviewer: ```suggestion
```

>   | US     | ingest.us.signoz.cloud:443 |

Reviewer: ```suggestion
```

>   | IN     | ingest.in.signoz.cloud:443 |

Reviewer: ```suggestion
```

>   | EU     | ingest.eu.signoz.cloud:443 |

Reviewer: ```suggestion
```

> - The value of `SIGNOZ_ACCESS_TOKEN` environment variable will be `<SIGNOZ_INGESTION_KEY>` where `<SIGNOZ_INGESTION_KEY>` is your ingestion key

Reviewer: ```suggestion
```

</TabItem>

<TabItem value="self-host" label="Self-Host" default>

> For sending logs to Self hosted Signoz, we need it's running instance. You can follow these steps to [install](https://signoz.io/docs/install/docker/) signoz and set the below environment variables

Reviewer: ```suggestion
For sending logs to Self hosted SigNoz, we need it's running instance. You can follow these steps to [install](https://signoz.io/docs/install/docker/) SigNoz and set the below environment variables
```

- The value of `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable will be `https://localhost:4318/v1/logs` and here we don't need `SIGNOZ_ACCESS_TOKEN`

```javascript
const otlpExporter = new OTLPTraceExporter({
  url: 'http://localhost:4318/v1/logs',
  headers: {
    'Content-Type': 'application/json',
  },
})
```

Set a proxy in your `package.json` pointing to to the SigNoz OTLP HTTP endpoint. This is an easier temporary setup to bypass CORS issues. You can use [this documentation](https://signoz.io/docs/userguide/otlp-http-enable-cors/) to guide you on dealing with CORS issues for a permanent setup.

```bash
{
      "proxy": "http://localhost:4318"
}
```

</TabItem>
</Tabs>

> 

Reviewer: ```suggestion
## Setup
```

> ## How to collect logs from Node.js Winston without the requiring Otel collectors

Reviewer: ```suggestion
```

> - First we will install a few dependencies using npm

Reviewer: ```suggestion
### Step 1: Install Dependencies
```
Give a couple of lines brief about what these libraries are and what they will be used for.

  ```
  npm install @opentelemetry/api-logs @opentelemetry/sdk-logs @opentelemetry/exporter-logs-otlp-http @opentelemetry/winston-transport @opentelemetry/resources winston dotenv
  ```

> - logger.js

Reviewer: ```suggestion
### Step 2: Create `logger.js` file
```
Explain what this file is for and how what the below code block is doing

>   ```javascript

Reviewer: ```suggestion
  ```javascript:logger.js
```

    }),
  })

  // Configure OTLP exporter for SigNoz
  const otlpExporter = new OTLPLogExporter({
    url: process.env.OTEL_EXPORTER_OTLP_ENDPOINT,
    headers: {
      'signoz-access-token': process.env.SIGNOZ_ACCESS_TOKEN,
    },
  })

  // Add processor with the OTLP exporter
  loggerProvider.addLogRecordProcessor(new SimpleLogRecordProcessor(otlpExporter))

  // Set the global logger provider
  logsAPI.logs.setGlobalLoggerProvider(loggerProvider)

  // Create Winston logger with Console and OpenTelemetry transports
  const logger = winston.createLogger({
    level: 'debug',
    format: winston.format.combine(
      winston.format.timestamp(),
      winston.format.errors({ stack: true }),
      winston.format.metadata(),
      winston.format.json()
    ),
    defaultMeta: {
      service: 'winston-logger',
      environment: process.env.NODE_ENV || 'development',
    },
    transports: [
      new OpenTelemetryTransportV3({
        loggerProvider,
        logAttributes: {
          'service.name': 'winston-logger',
          'deployment.environment': process.env.NODE_ENV || 'development',
        },
      }),
    ],
  })

  module.exports = logger
  ```

> - example index.js

Reviewer: ```suggestion
### Step 3: Import in `index.js` file
```
Please find heading for this step and bifurcate this code block into different code block for different methods as we have discussed.

      // Simulating an authentication failure
      throw new Error('Authentication Failed')
    } catch (error) {
      logger.error({
        message: 'User Authentication Error',
        error: error.message,
        stack: error.stack,
        requestId: req.requestId,
      })

      res.status(401).json({
        error: 'Unauthorized',
        requestId: req.requestId,
      })
    }
  })

  app.use((err, req, res, next) => {
    logger.error({
      message: 'Unhandled Error',
      error: err.message,
      stack: err.stack,
      requestId: req.requestId || 'unknown',
    })

    res.status(500).json({
      error: 'Unexpected Server Error',
      requestId: req.requestId || 'unknown',
    })
  })

  const server = app.listen(PORT, () => {
    logger.info({
      message: 'Server Started',
      port: PORT,
      environment: process.env.NODE_ENV || 'development',
    })

    console.log(`Listening on http://localhost:${PORT}`)
  })

  process.on('SIGTERM', () => {
    logger.warn('SIGTERM received. Shutting down gracefully.')
    server.close(() => {
      logger.info('Server closed. Process terminating.')
      process.exit(0)
    })
  })

  module.exports = app
  ```

> - Now you can run your application by running `node index.js`

Reviewer: ```suggestion
- #Step 4: Run application
```

> - If there are no errors your logs will be visible on SigNoz UI.

Reviewer: ```suggestion
If there are no errors your logs will be visible on SigNoz UI.
```
Put the output image below this

## Collecting Nodejs logs when application is deployed on Docker or Kubernetes

When your application is deployed in Docker or a Kubernetes cluster, the logs from the console are automatically collected and stored in the node. The SigNoz collector will automatically collect the logs and they will be visible on the SigNoz UI.

### Console Transport Configuration

You can easily add a console transport to your Winston logger with the following code:

> ```javascript

Reviewer: Mention where this codeblock will go and explain it with an example.

logger.add(new winston.transports.Console(options))
```

The Console transport supports several configuration options:

- **level:** Specifies the level of messages this transport should log (default: level set on parent logger).
- **silent:** A boolean flag to suppress output (default is false).
- **eol:** Defines the end-of-line characters to use (default is `os.EOL`).
- **stderrLevels:** An array of log levels to be sent to stderr instead of stdout. For example: `['error', 'debug', 'info']` (default is an empty array).
- **consoleWarnLevels:** An array of log levels to use `console.warn()` or stderr (in Node.js) instead of stdout. For example: `['warn', 'debug']` (default is an empty array).

Example configuration:

```javascript
logger.add(
  new winston.transports.Console({
    level: 'info',
    format: winston.format.simple(),
    silent: false,
    stderrLevels: ['error'],
  })
)
```

By using these configuration options, you can fine-tune how your logs are output and collected when deployed in containerized environments like Docker or Kubernetes.

> ## Collecting Nodejs logs when application is deployed on a Host

Reviewer: Remove this section

---
date: 2024-12-06
title: Collecting Nodejs winston logs
id: collecting_nodejs_winston_logs
---

## Overview

> In this documentation, you'll learn how to seamlessly export logs from your Node.js application to SigNoz using the `winston` logging library. `winston` provides various transports that make it easy to forward logs to external systems, and SigNoz offers you a platform for log monitoring and analysis.

Reviewer: ```suggestion
In this documentation, you'll learn how to seamlessly export logs from your Node.js application to SigNoz using the `winston` logging library. Winston provides various transports that make it easy to forward logs to external systems, and SigNoz offers you a platform for log monitoring and analysis.
```

> ## Prerequisites

Reviewer: ```suggestion
```
We can completely remove these Pre-Requisites.

### 1: For Signoz Cloud

- SigNoz Cloud Account: Create an account on SigNoz Cloud
- Ingestion Key: Obtain your ingestion key from the SigNoz Cloud dashboard

### 2: For Self-Hosted SigNoz

- Running Instance of SigNoz: Follow these steps to [install](https://signoz.io/docs/install/docker/) and run a self-hosted instance of SigNoz.

## Setup

### Step 1: Install Dependencies

```bash
  npm install @opentelemetry/api-logs @opentelemetry/sdk-logs @opentelemetry/exporter-logs-otlp-http @opentelemetry/winston-transport @opentelemetry/resources winston dotenv
```

### Step 2: Create `logger.js` file

> In the `src` directory of your web application, create a new JavaScript file (e.g., logger.js). The `logger.js` file is set up to provide the initial configuration for logging in your Node.js application. It creates a logger using `winston` and configures it to send logs to SigNoz for monitoring. You can use this logger as-is or customize it further based on your project’s requirements. Once set up, you can import and use it wherever logging is needed in your application.

Reviewer: ```suggestion
In the `src` directory of your web application, create a new JavaScript file `logger.js`. The `logger.js` file is set up to provide the initial configuration for logging in your Node.js application. It creates a logger using `winston` and configures it to send logs to SigNoz for monitoring. You can use this logger as-is or customize it further based on your project’s requirements. Once set up, you can import and use it wherever logging is needed in your application.
```

```javascript:logger.js
  const logsAPI = require('@opentelemetry/api-logs')
  const { LoggerProvider, SimpleLogRecordProcessor } = require('@opentelemetry/sdk-logs')
  const { OTLPLogExporter } = require('@opentelemetry/exporter-logs-otlp-http')
  const { OpenTelemetryTransportV3 } = require('@opentelemetry/winston-transport')
  const { Resource } = require('@opentelemetry/resources')
  const winston = require('winston')
  require('dotenv').config()
```

The `LoggerProvider` manages logs for your application.
`Resource`: Adds metadata about the application, such as service.name, service.version, and the current environment (development by default).

```javascript:logger.js
// Initialize the Logger provider
const loggerProvider = new LoggerProvider({
  resource: new Resource({
    'service.name': 'winston-logger',
    'service.version': '1.0.0',
    'deployment.environment': process.env.NODE_ENV || 'development',
  }),
})
```

> `OTLPLogExporter` defines the endpoint where logs are exported. It uses the environment variables `OTEL_EXPORTER_OTLP_ENDPOINT` and `SIGNOZ_ACCESS_TOKEN`

Reviewer: ```suggestion
`OTLPLogExporter` defines the endpoint where logs are exported. It uses the environment variables `OTEL_EXPORTER_OTLP_ENDPOINT` and `SIGNOZ_INGESTION_KEY`
```
The term Access token is not used now. We have renamed it to Ingestion key

```javascript:logger.js
// Configure OTLP exporter for SigNoz

const otlpExporter = new OTLPLogExporter({
  url: process.env.OTEL_EXPORTER_OTLP_ENDPOINT,
  headers: {
    'signoz-access-token': process.env.SIGNOZ_ACCESS_TOKEN,
  },
})
```

> <Tabs entityName="plans">

Reviewer: Let's move the tabs to above `logger.js` section since Ingestion key is not available in the self-host section.

Video explanation - [here](https://www.loom.com/share/a9c672f905734e32b9a9681bc4115ead?sid=1225151b-799b-4880-a751-74bd02c628bc)

<TabItem value="cloud" label="SigNoz Cloud" default>

For sending logs to SigNoz cloud, while running the above example set the below environment variables

> - The value of `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable will be `https://ingest.{region}.signoz.cloud:443/v1/logs` where depending on the choice of your region for SigNoz cloud, the otlp endpoint will vary according to this table.

Reviewer: Please create an environment file code block with the below content. This should be Step 2.

----
### Step 2: Create an Environment file

Create a `.env` file with the environment variables needed in the following steps.

```ini
NODE_ENV=<your-environment-name> #Ex. -> dev, prod etc.
OTEL_EXPORTER_OTLP_ENDPOINT= https://ingest.<region>.signoz.cloud:443/v1/logs
SIGNOZ_INGESTION_KEY=<your-ingestion-key>
```
- Set the `<region>` to match your SigNoz Cloud [region](https://signoz.io/docs/ingestion/signoz-cloud/overview/#endpoint)
- Replace `<your-ingestion-key>` with your SigNoz [ingestion key](https://signoz.io/docs/ingestion/signoz-cloud/keys/)

---

**Renumber the other steps since this will be the second step now.** Please create separate tabs for Cloud and Self-Host environment files. Self-Host one will not have the Ingestion Key Variable.

> - The value of `SIGNOZ_ACCESS_TOKEN` environment variable will be `<SIGNOZ_INGESTION_KEY>` where `<SIGNOZ_INGESTION_KEY>` is your ingestion key

Reviewer: ```suggestion
```

Should be taken care of when working on the above comment

</TabItem>

<TabItem value="self-host" label="Self-Host" default>
For sending logs to Self hosted Signoz, we need it's running instance. You can follow these steps to [install](https://signoz.io/docs/install/docker/) Signoz and set the below environment variables

> - The value of `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable will be `https://localhost:4318/v1/logs` and here we don't need `SIGNOZ_ACCESS_TOKEN`

Reviewer: ```suggestion
- The value of `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable will be `https://localhost:4318/v1/logs` and here we don't need `SIGNOZ_INGESTION_KEY`
```

// Set the global logger provider
logsAPI.logs.setGlobalLoggerProvider(loggerProvider)
```

Create Winston logger with OpenTelemetry transports. Here we used `OpenTelemetryTransportV3`, which sends logs to the global loggerProvider.

```javascript:logger.js
const logger = winston.createLogger({
    level: 'debug',
    format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.errors({
            stack: true
        }),
        winston.format.metadata(),
        winston.format.json()
    ),
    defaultMeta: {
        service: 'winston-logger',
        environment: process.env.NODE_ENV || 'development',
    },
    transports: [
        new OpenTelemetryTransportV3({
            loggerProvider,
            logAttributes: {
                'service.name': 'winston-logger',
                'deployment.environment': process.env.NODE_ENV || 'development',
            },
        }),
    ],
})

module.exports = logger

```

### Step 3: Send Node Winston logs to SigNoz

To send logs to SigNoz, import `logger.js` into `index.js` and save the changes.

```javascript:index.js
  const logger = require('./logger')
```

Logging scenarios

> 1. Request Logging Middleware - Track all incoming requests and their completion with detailed context:

Reviewer: ```suggestion
1. **Request Logging Middleware** - Track all incoming requests and their completion with detailed context:
```

```javascript:index.js
app.use((req, res, next) => {
  const requestId = uuidv4();
  req.requestId = requestId;

  logger.info({
    message: 'Incoming Request',
    method: req.method,
    path: req.path,
    requestId: requestId,
    ipAddress: req.ip,
    userAgent: req.get('User-Agent'),
  });

  req.startTime = Date.now();

  res.on('finish', () => {
    const duration = Date.now() - req.startTime;
    logger.info({
      message: 'Request Processed',
      method: req.method,
      path: req.path,
      requestId: requestId,
      statusCode: res.statusCode,
      responseTime: `${duration}ms`,
    });
  });

  next();
});
```

> 2. Debug Logging - Use debug logging during development for troubleshooting:

Reviewer: ```suggestion
2. **Debug Logging** - Use debug logging during development for troubleshooting:
```

```javascript:index.js
app.get('/', (req, res) => {
  try {
    logger.debug({
      message: 'Accessing root endpoint',
      requestId: req.requestId,
    });

    // Route logic...
  } catch (error) {
    logger.error({
      message: 'Error in root endpoint',
      error: error.message,
      stack: error.stack,
      requestId: req.requestId,
    });
    res.status(500).json({ error: 'Internal Server Error' });
  }
});
```

> 3. Warning Logging - Log non-critical issues that require attention:

Reviewer: ```suggestion
3. **Warning Logging** - Log non-critical issues that require attention:
```

```javascript:index.js
app.get('/404', (req, res) => {
  logger.warn({
    message: '404 Route Accessed',
    requestId: req.requestId,
  });

  res.status(404).json({
    error: 'Not Found',
    requestId: req.requestId,
  });
});
```

> 4. Error Logging - Capture and log critical errors with full context:

Reviewer: ```suggestion
4. **Error Logging** - Capture and log critical errors with full context:
```

```javascript:index.js
app.get('/user', (req, res) => {
  try {
    throw new Error('Authentication Failed');
  } catch (error) {
    logger.error({
      message: 'User Authentication Error',
      error: error.message,
      stack: error.stack,
      requestId: req.requestId,
    });

    res.status(401).json({
      error: 'Unauthorized',
      requestId: req.requestId,
    });
  }
});
```

> 5. Global Error Handler - Implement a catch-all error handler:

Reviewer: ```suggestion
5. **Global Error Handler** - Implement a catch-all error handler:
```

```javascript:index.js
app.use((err, req, res, next) => {
  logger.error({
    message: 'Unhandled Error',
    error: err.message,
    stack: err.stack,
    requestId: req.requestId || 'unknown',
  });

  res.status(500).json({
    error: 'Unexpected Server Error',
    requestId: req.requestId || 'unknown',
  });
});
```

> 6. Application Lifecycle Logging - Track application startup and shutdown events:

Reviewer: ```suggestion
6. **Application Lifecycle Logging** - Track application startup and shutdown events:
```

```javascript:index.js
const server = app.listen(PORT, () => {
  logger.info({
    message: 'Server Started',
    port: PORT,
    environment: process.env.NODE_ENV || 'development',
  });
});

process.on('SIGTERM', () => {
  logger.warn('SIGTERM received. Shutting down gracefully.');
  server.close(() => {
    logger.info('Server closed. Process terminating.');
    process.exit(0);
  });
});

```

### Step 4: Run your application

```bash
  npm start
```

<figure data-zoomable align="center">
  <img
    className="box-shadowed-image"
    src="/img/docs/logs-management/send-logs/node-winston-logs.webp"
    alt="Send Nodejs winston logs to Signoz"
  />
  <figcaption>
    <i>Send Nodejs winston logs to Signoz</i>
  </figcaption>
</figure>

> ## Collecting Nodejs logs when application is deployed on Docker or Kubernetes

Reviewer: ```suggestion
## Application is deployed on Docker or Kubernetes
```
